src/dune/sub_dirs.ml

```diff
@@ -91,23 +78,30 @@ let make ~dirs ~data_only ~ignored_sub_dirs ~vendored_dirs =
 
 let eval (t : _ Status.Map.t) ~dirs =
   let normal = Predicate_lang.filter t.normal ~standard:default.normal dirs in
-  let to_set ~standard pred =
-    String.Set.of_list (Predicate_lang.filter pred ~standard dirs)
+  let eval ~standard pred = Predicate_lang.filter pred ~standard dirs in
+  let data_only = eval ~standard:default.data_only t.data_only in
+  let vendored = eval ~standard:default.vendored t.vendored in
+  let statuses =
+    List.fold_left normal ~init:String.Map.empty
+      ~f:(fun acc dir -> String.Map.set acc dir Status.Normal)
   in
-  let data_only = to_set ~standard:default.data_only t.data_only in
-  let vendored = to_set ~standard:default.vendored t.vendored in
-  let both_vendored_and_data = String.Set.inter data_only vendored in
-  match String.Set.choose both_vendored_and_data with
-  | None ->
-    let normal = String.Set.of_list normal in
-    { Status.Map.normal; data_only; vendored }
-  | Some dir ->
-    User_error.raise
-      [ Pp.textf
-          "Directory %s was marked as vendored and data_only, it can't be \
-           marked as both."
-          dir
-      ]
+  let statuses =
+    List.fold_left data_only ~init:statuses ~f:(fun acc dir ->
+      String.Map.set acc dir Status.Data_only)
+  in
+  List.fold_left vendored ~init:statuses ~f:(fun acc dir ->
+    String.Map.update acc dir ~f:(function
+      | None
```

Comment:
Looks like this additionally changes what was previously a case of `false, false, true` from "Ignored" to "Vendored". That seems like a more expected behavior, but I guess it's not strictly backwards-compatible?

Reply:
Indeed, I think this is more consistent. I will update the change log to reflect this.